tests/js/track-timeouts.js

```diff
@@ -0,0 +1,74 @@
+/**
+ * Utilities for tracking timeouts.
+ *
+ * Site Kit by Google, Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Sets up global functions to track timeouts and wait for them to complete.
+ *
+ * This adds the global function `useTrackedTimeouts`, which will replace the
+ * global `setTimeout` and `clearTimeout` functions with versions that track
+ * the timeouts.
+ *
+ * `useTrackedTimeouts` also adds two global functions to the global scope:
+ * `hasTimeouts` and `waitForTimeouts`.
+ *
+ * `hasTimeouts` returns true if there are any timeouts currently active.
+ *
+ * `waitForTimeouts` returns a Promise that resolves when all timeouts have
+ * completed.
+ *
+ * @since n.e.x.t
+ *
+ * @return {void}
+ */
+export function setupTimeoutTracker() {
+	let originalTimeoutAPI;
+
+	global.useTrackedTimeouts = () => {
+		originalTimeoutAPI = originalTimeoutAPI || {
+			setTimeout: global.setTimeout,
+			clearTimeout: global.clearTimeout,
+		};
```

Comment:
I like the idea of this in principle but I'm not sure it's the right solution here. There are some concerns I have with this API and its coupling to whatever timer API is present at the time. Together, with the use of globals, this has the potential to be stacked on top of itself, as well as inconsistent results based on switching timer implementations with Jest.

Ideally, this would be something Jest would provide itself in order to provide a consistent experience with all timer implementations. From what I can tell, I don't think this is needed. See next comments.

Reply:
I understand your reticence here, I think some of your concerns could be alleviated with a bit more work, but ultimately we can do without it and as you say it's something that would be nice to see Jest provide itself...